### ISSUE TITLE
Consider debug mode for ArrayPool

### ISSUE THREAD

**Issue by danmoseley** [dotnet/runtime#7532]
There are several ways to misuse array pool

1. forget to return - merely inefficient
2. return and continue to use 
3. renter expects pre cleared array
4. returner does not request a clear and leaks sensitive data
5. double return

There is some logging to help trace suspected cases of 1 and 2.

For 4 if we start caring more about that someday we could change the default to clear data. 

For 5 we could at least (in a diagnostic mode) compare against existing arrays in the pool (perhaps an expanded size pool that returns LRU so it's likely still in the pool when returned the 2nd time)

We could help catch cases of both 2 and 3 by having a mode where we set the array to 0xDEADBEEF on return. We would do this whether or not the returner requested the array clear, because the next renter cannot rely on the returner.

This should be enabled in #if DEBUG but since running chk corelib is so rarely done, it might realistically only get used if it was #if DEBUG || DEBUG_ARRAYPOOL or something like that. So perhaps this isn't worth bothering with unless and until we have some investigation that it would help with.

Just a thought for the future.

**Comment by jkotas** [comment]
Related: https://developercommunity.visualstudio.com/content/problem/1027129/memory-corruption-when-pinning-arraypool-arrays.html

**Comment by danmosemsft** [comment]
Another option would be a different factory method to create the ArrayPool (perhaps overloads or instead of Shared() or Create()) that would return the diagnostic pool. Presumably there are relatively few places in one's code where the array pool is created (probably 1, if you know which array pool is leading to the problem) so changing that one line of code for testing purposes is not onerous. As a separate concrete implementation of ArrayPool performance would not be a concern and it could record callstacks, etc like a diagnostic heap.

**Comment by huoyaoyuan** [comment]
Can this be achieved with AppContext switch? Since the checks may be somehow intensive, I'm not sure about Tier-0 performance impact.

**Comment by kayle** [comment]
My team just spent a lot of time investigating several failures related to a `5. double return` error in an external component.

Has any perf analysis been done on what it would cost to always have a runtime check on returns to help catch this? A parity bit for pooled types that is 0 when in the pool and 1 when in use seems like it would catch most double-frees and have very low runtime overhead.

**Comment by tannergooding** [comment]
@EgorBo, @jkotas, @richlander 

This is likely worth considering as part of the "reducing unsafe" feature work or should otherwise be closed as "not planned".

The general concept (as was covered above) here is similar in premise to things like `memory leak detection` or `CRT debugging` provided by many native libraries. It could be "solved" (for both NAOT and JIT) via AppContext switches. 

Using app context switches for such validation is something I have done in `TerraFX.Interop.Mimalloc` (a managed port of the microsoft/mimalloc library) which provides multiple switches for additional security, debugging, and general correctness usage validation: https://github.com/terrafx/terrafx.interop.mimalloc/blob/main/sources/Interop/Mimalloc/Mimalloc.Init.cs#L137 (mirroring the `#define` macros supported by the C library).

This seems just as useful as things like the `GlobalizationMode.Invariant` switch we provide for NAOT and can allow easier debugging or safety for users who need it. The downside is that it cannot easily support dynamic toggling (there are mechanisms to achieve this and still allow trimming, however) and that the current API surface isn't exactly suited to tracking some of the initial callouts, like ensuring buffers are returned.

**Comment by EgorBo** [comment]
cc @stephentoub as someone who tried to address it via https://github.com/dotnet/runtime/pull/109695

**Comment by richlander** [comment]
> C# does not have a system for tracking "ownership" and "lifetime" for heap-allocated objects that can lead use-after-free violations, for example with ArrayPool<T>.Shared.Rent and Return.

Source: https://github.com/dotnet/designs/blob/main/accepted/2025/memory-safety/memory-safety.md

Yes, this is considered in-scope. No suggestion that ownership is the only mitigation to apply here.

**Comment by kayle** [comment]
Are there active efforts for more robust ownership tracking? If not, a single parity bit seems like a good place to start assuming it's relatively cheap to implement and performant enough to be enabled at runtime by default.

For my scenario, https://github.com/dotnet/runtime/pull/109695 wouldn't have helped much since it was an extension causing the corruption and not something we would've caught pro-actively.

**Comment by stephentoub** [comment]
> a single parity bit

A bit where?

**Comment by kayle** [comment]
> A bit where?

I don't know the details on that part :) Is there one available in the object header/SyncBlock space? Otherwise, for pooled arrays could you allocate extra storage after the end of the array? I guess you'd still need a bit to tag the array as having come from the pool in the first place which could then just as easily be the parity bit.

**Comment by tannergooding** [comment]
There is "free space" in the 64-bit array (the padding bits) and I imagine a bit could be found in the object header if really needed.

I don't think tracking whether it is in the pool or not via a bit is necessary, though. We could just check if the pool already contains the same array, particularly since we already bucket and the number of entries is expected to be small for the way `ArrayPool<T>.Shared` is setup.

Given that we're only supposed to be returning arrays that were rented and not arbitrary arrays, some kind of side table would also be sufficient to validate both that and the double return case.

**Comment by stephentoub** [comment]
> There is "free space" in the 64-bit array (the padding bits) and I imagine a bit could be found in the object header if really needed.
> 
> I don't think tracking whether it is in the pool or not via a bit is necessary, though. We could just check if the pool already contains the same array, particularly since we already bucket and the number of entries is expected to be small for the way `ArrayPool<T>.Shared` is setup.
> 
> Given that we're only supposed to be returning arrays that were rented and not arbitrary arrays, some kind of side table would also be sufficient to validate both that and the double return case.

Last I tried, those had larger overheads that we would want enabled by default. Why is why https://github.com/dotnet/runtime/pull/109695 did it as opt-in.

**Comment by tannergooding** [comment]
Definitely agree that we probably don't want it on by default. I don't think any approach would reasonably allow for that.

However, I do think an opt-in switch would provide some benefit and there may be some explicit "validation" APIs that could be exposed for users who want to be more explicit about their safety or add their own `Debug.Assert` checks.

For example, we could have some `ArrayPool<T>.Contains(T[])` method which checks if the array is already in the pool. Users wanting additional validation can then call it. Some opt-in switch would call it as part of `Return(T[])`. Doing some `ArrayPool<T>.IsCompatible(T[])` (I don't like that name) that checks if it is an array explicitly created by that pool is harder to do without an opt-in, but there's likely something we can do there too without significant cost.

**Comment by kayle** [comment]
>Last I tried, those had larger overheads that we would want enabled by default. Why is why https://github.com/dotnet/runtime/pull/109695 did it as opt-in.

Do you expect the parity bit approach in an object header to also have a measureable perf impact? Or only that walking the possible items in the pool is too much?

**Comment by jkotas** [comment]
Also from https://github.com/dotnet/designs/blob/main/accepted/2025/memory-safety/memory-safety.md:

> Potential additional tools that look for specific bad patterns in unsafe code will be a lower priority. Developers writing and reviewing unsafe code take on the sole responsibility for ensuring that the code is sound.

The parity bit idea is one of many potential tools like that. The challenge with building such tools is that the effort is expensive, endless, and does not actually fix the problem. C/C++ has spent incredible amount on energy on it, and memory safety vulnerabilities in C/C++ are still very prevalent.

Thus our plan is look into solutions that are safe by construction first. We may end up doing something here, but I expect that it will be seriously looked into only after we fully exploit safe by construction solutions first.
